mat/tridiag.go

```diff
@@ -0,0 +1,371 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack/lapack64"
+)
+
+var (
+	tridiagDense *Tridiag
+	_            Matrix           = tridiagDense
+	_            allMatrix        = tridiagDense
+	_            denseMatrix      = tridiagDense
+	_            Banded           = tridiagDense
+	_            MutableBanded    = tridiagDense
+	_            RawTridiagonaler = tridiagDense
+)
+
+// A RawTridiagonaler can return a lapack64.Tridiagonal representation of the
+// receiver. Changes to the DL, D, DU slices in lapack64.Tridiagonal will be
+// reflected in the original matrix, changes to the N field will not.
+type RawTridiagonaler interface {
+	RawTridiagonal() lapack64.Tridiagonal
+}
+
+// Tridiag represents a tridiagonal matrix by its three diagonals.
+type Tridiag struct {
+	mat lapack64.Tridiagonal
+}
+
+// NewTridiag creates a new n×n tridiagonal matrix with the first sub-diagonal
+// in dl, the main diagonal in d and the first super-diagonal in du. If dl, d,
+// and du are nil, new backing slices will be allocated for them. If dl and du
+// have length n-1 (or are nil if n is 1) and d has length n, they will be used
+// as backing slices, and changes to the elements of the returned Tridiag will
+// be reflected in dl, d, du. If neither of these is true, NewTridiag will
+// panic.
+func NewTridiag(n int, dl, d, du []float64) *Tridiag {
+	switch {
+	case n <= 0:
+		if n == 0 {
+			panic(ErrZeroLength)
+		}
+		panic(ErrNegativeDimension)
+	case n == 1:
+		if dl != nil || du != nil || (d != nil && len(d) != 1) {
+			panic(ErrShape)
+		}
+		if d == nil {
+			d = make([]float64, 1)
+		}
+	default:
+		if dl != nil || d != nil || du != nil {
+			if len(dl) != n-1 || len(d) != n || len(du) != n-1 {
+				panic(ErrShape)
+			}
+		} else {
+			dl = make([]float64, n-1)
+			d = make([]float64, n)
+			du = make([]float64, n-1)
+		}
+	}
+	return &Tridiag{
+		mat: lapack64.Tridiagonal{
+			N:  n,
+			DL: dl,
+			D:  d,
+			DU: du,
+		},
+	}
+}
+
+// Dims returns the number of rows and columns in the matrix.
+func (a *Tridiag) Dims() (r, c int) {
+	return a.mat.N, a.mat.N
+}
+
+// Bandwidth returns 1, 1 - the upper and lower bandwidths of the matrix.
+func (a *Tridiag) Bandwidth() (kl, ku int) {
+	return 1, 1
+}
+
+// T performs an implicit transpose by returning the receiver inside a Transpose.
+func (a *Tridiag) T() Matrix {
+	// TODO(vladimir-ch): should we return the receiver with DL,DU swapped?
```

Comment:
I guess this depends on how backing routines handle it. From memory it's just a swapping of diags rather than an trans param, no?

Reply:
`Dlagtm` (and `Lagtm`) take the trans parameter, `Dgtsv` does not but we've added it to `Gtsv` for consistency and so that the caller doesn't have to fiddle with the fields of `lapack64.Triangular`. But I think that doesn't affect the choice here. With swapping `untransposeExtract` will just always return false, so in the end there's not much difference. Maybe we should keep using the Transpose type for consistency unless we want to take advantage of the easy transposing that Tridiag representation offers.